index/segment/mem/postings_list.go

```diff
@@ -0,0 +1,171 @@
+
+package mem
+
+import (
+	"math"
+	"sync"
+
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+type postingID int
+
+// postingsList represents an efficient mapping from postingID -> bitmap of doc ids
+type postingsList interface {
+	// InsertNew inserts document `i` at a new posting id, and returns the
+	// positing id used.
+	InsertNew(i segment.DocID) postingID
+
+	// Insert inserts a reference to document `i` for posting id `p`.
+	Insert(p postingID, i segment.DocID)
+
+	// Remove removes the reference (if any) from document `i` from posting id `p`.
+	Remove(p postingID, i segment.DocID)
+
+	// Fetch retrieves all documents which are associated with posting id `p`.
+	Fetch(p postingID) *docIDs
+}
+
+type pList struct {
+	sync.RWMutex
+	// TODO(prateek): test perf impact of slice v map here
+	postings []*docIDs
+
+	counter *atomic.Int32
+}
+
+func newPostingsList(initialSize int) postingsList {
+	return &pList{
+		postings: make([]*docIDs, initialSize),
+		counter:  atomic.NewInt32(-1),
+	}
+}
+
+func (p *pList) InsertNew(i segment.DocID) postingID {
+	t := postingID(p.counter.Inc())
+	p.Insert(t, i)
+	return t
+}
+
+func (p *pList) Insert(t postingID, i segment.DocID) {
+	p.Lock()
+	// ensure postings has enough capacity to store the document
+	if int(t) >= len(p.postings) {
+		// amortized O(1) expansion
+		factor := math.Ceil(float64(t) / float64(len(p.postings)))
+		if factor < 2 {
+			factor = 2
+		}
+		newPostings := make([]*docIDs, 1+int(factor)*len(p.postings))
+		copy(newPostings, p.postings)
+		p.postings = newPostings
+	}
+
+	// allocate bitmap if undeclared
+	ids := p.postings[t]
+	if ids == nil {
+		ids = newDocIDs()
+		p.postings[t] = ids
+	}
+	p.Unlock()
+
+	// set doc id
+	ids.insert(i)
+}
+
+func (p *pList) Remove(t postingID, i segment.DocID) {
+	p.Lock()
+
+	// terminate early if term offset is beyond capacity
+	if int(t) >= len(p.postings) {
+		p.Unlock()
+		return
+	}
+
+	ids := p.postings[t]
+	p.Unlock()
+
+	// terminate early if we don't have a bitmap for the postingID
+	if ids == nil {
+		return
+	}
+
+	ids.remove(i)
+}
+
+func (p *pList) Fetch(t postingID) *docIDs {
+	p.RLock()
+	if int(t) >= len(p.postings) {
+		p.RUnlock()
+		return nil
+	}
+
+	ids := p.postings[t]
+	p.RUnlock()
+	return ids
+}
+
+type docIDs struct {
+	sync.RWMutex
+	bitmap *roaring.Bitmap
+}
+
+func newDocIDs() *docIDs {
+	return &docIDs{
+		bitmap: roaring.NewBitmap(),
+	}
+}
+
+func (d *docIDs) insert(i segment.DocID) {
+	d.Lock()
+	d.bitmap.Add(uint32(i))
+	d.Unlock()
+}
+
+func (d *docIDs) remove(i segment.DocID) {
+	d.Lock()
+	d.bitmap.Remove(uint32(i))
+	d.Unlock()
+}
+
+func (d *docIDs) contains(i segment.DocID) bool {
+	d.RLock()
+	contains := d.bitmap.Contains(uint32(i))
+	d.RUnlock()
+	return contains
+}
+
+func (d *docIDs) clone() *roaring.Bitmap {
+	d.RLock()
+	clone := d.bitmap.Clone()
+	d.RUnlock()
+	return clone
+}
+
+func (d *docIDs) andTo(b *roaring.Bitmap) {
```

Comment:
`intersect` seems clearer to me, what do you think?

Reply:
`intersect` is definitely better than `andTo`. I don't like that it implies the underlying object `d *docIDs` is being changed. How would you feel about `project` or `intersectWith`

Reply:
I'd prefer `intersectWith` personally, `project` seems a little too vague.

Reply:
sold, made the change